### DISCUSSION TITLE
How to refresh app post deploy

### DISCUSSION THREAD

**Discussion by chris-carrington** [sveltejs/kit#8665] [Q&A] [answered]
Thank you for building SvelteKit!

I have my prod site open in a tab, push to main, vercel see's the push and deploys to production and vercel tells me all done in 15-20 seconds, amazing! 

I click on the opened prod site tab, click a link (client side routing) to the view with new content & it is not there, a page refresh shows the new content.

Goal: If app the user is viewing is not the latest have links refresh the page!

[I found this that may help,](https://github.com/sveltejs/kit/issues/546) idk how to know if current app version is not latest tho, thank you!

**Comment by teemingc** [comment]
Is it possible that your `app.html` or some other part of your app is preloading stale code / data?

https://kit.svelte.dev/docs/link-options#data-sveltekit-preload-data

https://kit.svelte.dev/docs/modules#$app-navigation-preloadcode

Otherwise, you could disable client-side routing by forcing a reload on all links if your app will be deployed frequently

https://kit.svelte.dev/docs/link-options#data-sveltekit-reload

**Comment by geoffrich** [comment, marked as answer]
You should be able to set [`version.pollInterval`](https://kit.svelte.dev/docs/configuration#version) in your svelte.config.js. With this setting, SvelteKit will check for a new version of your app using the interval you set. For example, with the below config, it will check every 5 seconds (5000 ms).

```js
const config = {
	kit: {
		adapter: adapter(),
		version: {
			pollInterval: 5000
		}
	}
};
```

If you've set `pollInterval` to a non-zero value, you can use the [`updated` store](https://kit.svelte.dev/docs/modules#$app-stores-updated) to determine if a new version is available.

```html
<script>
	import { updated } from '$app/stores';
</script>

<p>Updated: {$updated}</p>
```

If you want to force a reload whenever the app code has been updated, you can use [`data-sveltekit-reload`](https://kit.svelte.dev/docs/link-options#data-sveltekit-reload) like so in your root `+layout.svelte`:

```html
<script>
	import Header from './Header.svelte';
	import './styles.css';
	import { updated } from '$app/stores';
</script>

<div class="app" data-sveltekit-reload={$updated ? "" : "off"}>
	<slot />
</div>
```

When `$updated` is true, any navigation will do a full page reload. Otherwise when `$updated` is false, navigations will happen client side.

**Reply by teemingc** [reply]
The docs also state that SvelteKit handles this automatically by checking `version.name`. I wonder why that's not happening in this case.

> Client-side navigation can be buggy if you deploy a new version of your app while people are using it. If the code for the new page is already loaded, it may have stale content; if it isn't, the app's route manifest may point to a JavaScript file that no longer exists. SvelteKit solves this problem by falling back to traditional full-page navigation if it detects that a new version has been deployed, using the name specified here (which defaults to a timestamp of the build).

https://kit.svelte.dev/docs/configuration#version

**Reply by chris-carrington** [reply]
Thank you @geoffrich ! This looks lovely but I haven't gotten it to work, have more details, hopefully we're close!

1. [I added the code around my slot in layout](https://github.com/chris-carrington/sense-then-love/blob/main/src/routes/%2Blayout.svelte)
2. [I added the the code to svelte config](https://github.com/chris-carrington/sense-then-love/blob/main/svelte.config.js)
3. I deployed then refreshed the page
4. I deployed again then refreshed the page
5. I deployed to the **Contact Page**, the removal of an exclamation point @ the heading **Send us an email?!**
6. Deploy finishes I click in nav between Home & Contact, I do not see the removal of the exclamation point & [no request to _page.svelte.js in the Network tab](https://raw.githubusercontent.com/chris-carrington/sense-then-love/main/src/lib/images/deploy/no-works.jpg)
7. I do a browser reload on the Home page, click Contact in nav & I see the removal of the exclamation point & [a request to _page.svelte.js in the Network tab](https://raw.githubusercontent.com/chris-carrington/sense-then-love/main/src/lib/images/deploy/works.jpg)

Thank you so much!

**Reply by chris-carrington** [reply]
Thank you @s3812497 ! That link is helpful! I have seen it work sometimes, but it's not worked more often! lol I think the issue is not calling _page.svelte in the stale case but that's some under the hood stuff that idk about!

**Reply by geoffrich** [reply]
@chris-carrington I believe your issue is that your Nav is outside of the element with `data-sveltekit-reload` - only links that are descendants of that element will have the full-page-refresh behavior applied. Because your nav links are outside of the `main` element with the `data-sveltekit-reload` attribute, they won't inherit the behavior. So you might need to do something like this:

```diff
+ <div data-sveltekit-reload={ $updated ? '' : 'off' }>
  <main>
    <slot />
  </main>
  <Nav />
  <Background />
  <NavigationComplete />
  <div id="toast-wrapper"></div>
+ </div>
```

@s3812497 ah, thanks for pointing that out - I thought something like that existed but skipped right over that part of the docs. If someone can put together a minimal repro of it not working correctly we can investigate further (or maybe a docs update is needed instead to clarify things).

FWIW I was digging around earlier and I think the docs are referring to this code, but not sure which cases it covers:

https://github.com/sveltejs/kit/blob/4a44e673ef4bda56f89a103d7330a7f6815f12cd/packages/kit/src/runtime/client/client.js#L290-L295

**Reply by chris-carrington** [reply]
Thank you @geoffrich ! It's not working yet, just gave it two tests, the clicks stay client side w/ no refresh, here is my [layout](https://github.com/chris-carrington/sense-then-love/blob/main/src/routes/%2Blayout.svelte), & here is my [site](https://sensethenlove.com/), thank you so much!

**Reply by teemingc** [reply]
I've created an issue for this #8697
The `updated` store does not update consistently at the moment.

**Reply by chris-carrington** [reply]
Thank you @s3812497 ! I am sory but I am not following how that issue + answer relates to this discussion, do I also need to make additions to `svelte.config.js` to get this to work please?!

**Reply by teemingc** [reply]
@chris-carrington sorry about that, I had the wrong issue linked. Here it is #8697
Simon has also introduced an example solution in this comment https://github.com/sveltejs/kit/issues/8697#issuecomment-1403368458

**Reply by geoffrich** [reply]
@chris-carrington I see the issue -- you have the `version` config in the wrong place. It needs to be under `kit`

https://github.com/chris-carrington/sense-then-love/blob/52011fc73be23b14187bab143703cc77ec559985/svelte.config.js

```js
const config = {
  preprocess: vitePreprocess(),
  kit: {
    adapter: adapter(),
    version: {
      pollInterval: 5000
    }
  }
}
```

**Reply by chris-carrington** [reply]
@geoffrich Perfect! This worked! Thank you so much!!!

**Reply by whishkid** [reply]
Does this also work with adapter-static?

**Reply by teemingc** [reply]
yes. the version is a static file s the client side js can make a request to check it

**Reply by whishkid** [reply]
Thx!

**Reply by krishnaTORQUE** [reply]
> You should be able to set [`version.pollInterval`](https://kit.svelte.dev/docs/configuration#version) in your svelte.config.js. With this setting, SvelteKit will check for a new version of your app using the interval you set. For example, with the below config, it will check every 5 seconds (5000 ms).
> 
> ```js
> const config = {
> 	kit: {
> 		adapter: adapter(),
> 		version: {
> 			pollInterval: 5000
> 		}
> 	}
> };
> ```
> 
> If you've set `pollInterval` to a non-zero value, you can use the [`updated` store](https://kit.svelte.dev/docs/modules#$app-stores-updated) to determine if a new version is available.
> 
> ```
> <script>
> 	import { updated } from '$app/stores';
> </script>
> 
> <p>Updated: {$updated}</p>
> ```
> 
> If you want to force a reload whenever the app code has been updated, you can use [`data-sveltekit-reload`](https://kit.svelte.dev/docs/link-options#data-sveltekit-reload) like so in your root `+layout.svelte`:
> 
> ```
> <script>
> 	import Header from './Header.svelte';
> 	import './styles.css';
> 	import { updated } from '$app/stores';
> </script>
> 
> <div class="app" data-sveltekit-reload={$updated ? "" : "off"}>
> 	<slot />
> </div>
> ```
> 
> When `$updated` is true, any navigation will do a full page reload. Otherwise when `$updated` is false, navigations will happen client side.

`updated` has been deprecard. Is there any other alternative with the same behavior? (Clicking the link will do full page navigation or client side navigation). Thanks

**Reply by david-plugge** [reply]
https://svelte.dev/tutorial/kit/updated-state

**Comment by krishnaTORQUE** [comment]
@david-plugge Thanks a lot.
